### PR TITLE
UI: set memory max for oversubscription.

### PR DESCRIPTION
**Pull request by notnoop**
Display reserved vs max memory for allocations/tasks with memory oversubscription, introduced in https://github.com/hashicorp/nomad/pull/10247 .

Implements the alloc/task display memory info part of https://github.com/hashicorp/nomad/issues/10268 .

Needless to say, I very much lack javascript experience, and I'm unfamiliar with frontend patterns, so feel free to ignore the PR and do it the "proper" way and I can learn by watching it.  Thanks!

For a task with `memory_max` set, here are the screen shots:
Screen Shot 2021-04-06 at 12 09 09 PM

After:

Screen Shot 2021-04-06 at 12 10 29 PM

### CODE REVIEW

File: ui/tests/integration/components/primary-metric/allocation-test.js [hunk 1]
```diff
@@ -63,6 +63,22 @@ module('Integration | Component | PrimaryMetric::Allocation', function(hooks) {
...
+      groupCount: 1,
+      resourceSpec: ['M: 256, MM: 512'],
+    });
+    const alloc = this.server.create('allocation');
+
+    this.setProperties({ alloc, metric: 'memory' });
+    await render(template);
+
+    console.log(find('.annotation').innerText);
+    assert.ok(find('.annotation').innerText.includes('/ 256 MiB reserved, 512 MiB max'));
```
Comment (notnoop) [inline ui/tests/integration/components/primary-metric/allocation-test.js:79]: I've struggled with getting this test to pass - I always get `0 MiB / 0 MiB reserved`.  I get the same text in the above test as well.  Would love pointers for how to set it up.

File: ui/tests/unit/utils/allocation-stats-tracker-test.js [hunk 2]
```diff
@@ -189,13 +189,21 @@ module('Unit | Util | AllocationStatsTracker', function() {
...
+    console.log(JSON.stringify(expected) == JSON.stringify(tracker.get('tasks')));
+
     assert.deepEqual(
       tracker.get('tasks'),
-      [
-        { task: 'service', reservedCPU: 100, reservedMemory: 256, cpu: [], memory: [] },
-        { task: 'sidecar', reservedCPU: 50, reservedMemory: 128, cpu: [], memory: [] },
-        { task: 'log-shipper', reservedCPU: 50, reservedMemory: 128, cpu: [], memory: [] },
-      ],
+      expected,
```
Comment (notnoop) [inline ui/tests/unit/utils/allocation-stats-tracker-test.js:206]: This test is failing with cryptic message, and I'm unable to tell how the expected value differs from actual value. The json format for both values seem to be the same:

```
not ok 971 Chrome 89.0 - [15 ms] - Unit | Util | AllocationStatsTracker: append appropriately maps a data frame to the tracked stats for cpu and memory for the allocation as well as individual tasks
    ---
        actual: >
            [object Object],[object Object],[object Object]
        expected: >
            [object Object],[object Object],[object Object]
        stack: >
                at Object.<anonymous> (http://localhost:7357/ui/assets/tests.js:21918:14)
                at runTest (http://localhost:7357/ui/assets/test-support.js:27932:36)
                at Test.run (http://localhost:7357/ui/assets/test-support.js:27920:8)
                at http://localhost:7357/ui/assets/test-support.js:28142:15
                at processTaskQueue (http://localhost:7357/ui/assets/test-support.js:27530:32)
                at http://localhost:7357/ui/assets/test-support.js:27534:12
        message: >
            tasks represents the tasks for the allocation with no stats yet
        negative: >
            false
        browser log: |
            LOG: found
            LOG: [{"task":"service","reservedCPU":100,"reservedMemory":256,"cpu":[],"memory":[]},{"task":"sidecar","reservedCPU":50,"reservedMemory":128,"cpu":[],"memory":[]},{"task":"log-shipper","reservedCPU":50,"reservedMemory":128,"cpu":[],"memory":[]}]
            LOG: expected
            LOG: [{"task":"service","reservedCPU":100,"reservedMemory":256,"cpu":[],"memory":[]},{"task":"sidecar","reservedCPU":50,"reservedMemory":128,"cpu":[],"memory":[]},{"task":"log-shipper","reservedCPU":50,"reservedMemory":128,"cpu":[],"memory":[]}]
            LOG: true
```

from: https://app.circleci.com/pipelines/github/hashicorp/nomad/15529/workflows/ea9a55cd-cdcd-4fa0-83e0-995fc61e2adf/jobs/147809/steps

Comment (notnoop) [reply ui/tests/unit/utils/allocation-stats-tracker-test.js]: Apparently, the issue was that reservedMemoryMax was marked as `undefined` and json marshaller was skipping that field.

File: ui/tests/integration/components/primary-metric/allocation-test.js (hunk 1 above)
Comment (notnoop) [reply ui/tests/integration/components/primary-metric/allocation-test.js]: Apparently, the issue was a bit gnarlier than that - the tracker spy/mock didn't actually fill-in the annotation info, and there is something odd about the mirage creation if multiple jobs/annotations are in play.  Anyways, I took out the test, and would love to have a frontend engineer take over if possible :( sorry!

**Comment by notnoop** [comment]
Closing, superseded by https://github.com/hashicorp/nomad/pull/10459 .

**Comment by backspace** [comment]
#10459 isn’t quite the same as this, I just wanted to start with the bare minimum given the time constraints 😳

**Pull request closed without merging**